src/utils/__tests__/api-client.js

```diff
@@ -14,7 +14,12 @@ const defaultConfig = {
 
 const defaultResponse = {ok: true, json: () => Promise.resolve()}
 
+beforeEach(() => {
+  jest.spyOn(window, 'fetch')
```

Comment:
I'm confused how this could possibly work when we mock out `window.fetch` and it does nothing. I dove into the source of https://github.com/mswjs/node-request-interceptor and found that it doesn't reference `window.fetch` at all. How does this work when *my* code calls `window.fetch`, but that's been mocked to be a no-op 🤔

Reply:
Another strange thing is that if I move these hooks to `setupTests.js` the tests fail. I have to leave these hooks in here. That's very odd. I'm guessing it's a race condition...

Reply:
I've left this Jest spy because the assertions below call `window.fetch.mockResolvedValueOnce`, which would only exist if you've spied on `fetch`. This line is not needed to run mocks in Node, but was left for backwards-compatibility with the existing tests.

Reply:
Since `fetch` doesn't exist in Node, there is no direct reference to it in node-request-interceptor. Tests may still, however, reference fetch via `whatwg-fetch`/`node-fetch`/etc. polyfilled to be able to unit test a client-side code. Those polyfills either use `XMLHttpRequest` or `http`/`https` modules, which _are_ monkey patched in requests interceptor. 

So when your code calls `fetch` in Jest tests it most likely calls XHR (assuming Create React App), which is handled by the interception library.

Reply:
Now to why moving this spy to the global teardown fails the test, I suppose it may have something to do with how/when Jest establishes spies? Please, is there any detailed stack trace of what fails when this spy is moved out from here?

(I'd generally not move it from here, as this spy is specific to the assertions below, and no other code needs `fetch` to be spied on. Seeing this in a global tests teardown may me misleading).

Reply:
Figured out what was going on here. I mixed up some files. Moved everything to the `setupTests.js` and it's much cleaner :)

Reply:
Oh, and I also figured out how MSW manages to intercept `window.fetch` calls so that's awesome.

I'm curious whether you have suggestions on how to mock failed requests. Right now I'm providing a mock implementation of `window.fetch` for those, but I have a feeling it'd be better to change the handler for my MSW server. I'd want to be able to do that *within* the test itself though (rather than have to add a special handler in a separate file).

Reply:
Glad that worked nicely! 

As of now, I mock errors by returning a conditional response in my request handler. That implies a binding between the request and failed response. For example:

```js
rest.get('/book', (req, res, ctx) => {
  const bookId = req.url.searchParams.get('id')

  if (bookId === 'non-existing') {
    return res(
      ctx.status(404),
      ctx.text(`Book with id "${bookId}" not found`),
    )
  }

  return res(/* successful response */)
})
```

In the future I'd like to explore an API to alter request handlers at runtime (https://github.com/mswjs/msw/issues/128). Haven't though about a good API to do that, and what it would mean to the library's internals. 

Does that conditional response work for your use case?